src/globus_sdk/experimental/globus_app/authorizer_factory.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import annotations
+
+import abc
+
+from globus_sdk import AuthLoginClient, ConfidentialAppAuthClient
+from globus_sdk.authorizers import (
+    AccessTokenAuthorizer,
+    ClientCredentialsAuthorizer,
+    GlobusAuthorizer,
+    RefreshTokenAuthorizer,
+)
+from globus_sdk.experimental.tokenstorage import TokenData
+from globus_sdk.services.auth import OAuthTokenResponse
+
+from ._validating_token_storage import ValidatingTokenStorage
+from .errors import MissingTokensError
+
+
+class AuthorizerFactory(metaclass=abc.ABCMeta):
+    """
+    An ``AuthorizerFactory`` is an interface for getting some class of
+    ``GlobusAuthorizer`` from a ``ValidatingTokenStorage`` that meets the
+    authorization requirements used to initialize the ``ValidatingTokenStorage``.
+
+    An ``AuthorizerFactory`` keeps a cache of authorizer objects that are
+    re-used until its ``store_token_response`` method is called.
+    """
+
+    def __init__(self, token_storage: ValidatingTokenStorage):
+        """
+        :param token_storage: The ``ValidatingTokenStorage`` used
+        for defining and validating the set of authorization requirements that
+        constructed authorizers will meet and accessing underlying token storage
+        """
+        self.token_storage = token_storage
+        self._authorizer_cache: dict[str, GlobusAuthorizer] = {}
+
+    def _get_token_data_or_error(self, resource_server: str) -> TokenData:
+        token_data = self.token_storage.get_token_data(resource_server)
+        if token_data is None:
+            raise MissingTokensError(f"No token data for {resource_server}")
+
+        return token_data
+
+    def store_token_response(self, token_res: OAuthTokenResponse) -> None:
+        """
+        Store a token response in the underlying ``ValidatingTokenStorage``
+        and clear the authorizer cache.
+
+        :param token_res: An ``OAuthTokenResponse`` containing token data to be stored
+            in the underlying ``ValidatingTokenStorage``.
+        """
+        self.token_storage.store_token_response(token_res)
+        self._authorizer_cache = {}
+
+    def get_authorizer(self, resource_server: str) -> GlobusAuthorizer:
+        """
+        Either retrieve a cached authorizer for the given resource server or construct
+        a new one if none is cached.
+
+        Raises ``MissingTokensError`` if the underlying ``TokenStorage`` does not
+        have the needed tokens to create the authorizer.
+
+        :param resource_server: The resource server the authorizer will produce
+            authentication for
+        """
+        if resource_server in self._authorizer_cache:
+            return self._authorizer_cache[resource_server]
+
+        new_authorizer = self._make_authorizer(resource_server)
+        self._authorizer_cache[resource_server] = new_authorizer
+        return new_authorizer
+
+    @abc.abstractmethod
+    def _make_authorizer(self, resource_server: str) -> GlobusAuthorizer:
```

Comment:
I don't think it's essential, but this interface being the abstract part makes me think that this should be a generic class wherein `GlobusAuthorizer` is replaced with a bounded type var.

The construction would come out something like...
```python
GA = t.TypeVar("GA", bound=GlobusAuthorizer)

class AuthorizerFactory(t.Generic[GA]):
    def get_authorizer(self, resource_server: str) -> GA: ...
    def _make_authorizer(self, resource_server: str) -> GA: ...

class AccessTokenAuthorizerFactory(AuthorizerFactory[AccessTokenAuthorizer]): ...
```

I don't think it makes a big difference because the authorizer classes provide a uniform interface, but I think it's a closer match for your intent, in which a subclass is declared for each subtype.

Reply:
Yeah that's more what I was going for, thanks!